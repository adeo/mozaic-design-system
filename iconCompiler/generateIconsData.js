const dirtree = require('directory-tree')
const fs = require('fs')

const { iconsOutputPath } = require('./config')
const { createComponentName } = require('./tools')

const generateIconsData = () => {
  const iconDir = dirtree(iconsOutputPath.svg)
  const iconList = iconDir.children.map(icon => {
    const cat = icon.name.split('_')[0]
    const size = icon.name
      .replace(icon.extension, '')
      .split('_')
      .pop()

    const name = icon.name
      .replace(icon.extension, '')
      .replace(cat, '')
      .replace(size, '')
      .slice(1, -1)

    const fileName = icon.name

    const componentName = createComponentName(icon.name)

    return {
      cat,
      size,
      name,
      fileName,
      componentName,
    }
  })

  if (!fs.existsSync(`${process.cwd()}/src/data/`)) {
    fs.mkdirSync(`${process.cwd()}/src/data/`)
  }
  fs.writeFileSync(
    `${process.cwd()}/src/data/icons.json`,
    JSON.stringify(iconList, 0, 2),
    'utf8'
  )

  return true
}

module.exports = generateIconsData
