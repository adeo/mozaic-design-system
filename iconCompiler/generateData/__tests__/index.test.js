import fs from 'fs'

import generateIconsData from '../index'
import {
  getInfosFromName,
  generateMonochromListObject,
  generateColorListObject,
  checkPath,
  writeFile,
} from '../lib'

const iconList = [
  {
    data: `<svg></svg>`,
    fileName: 'Navigation_Arrow_Arrow--Top_16px.svg',
    outputPath: 'toto',
  },
  {
    data: `<svg></svg>`,
    fileName: 'Media_FullScreen_24px.svg',
    outputPath: 'toto',
  },
]

describe('getInfosFromName', () => {
  it('getsInfosFromName detect category, size, name, and filename from file name', () => {
    const data = getInfosFromName('Media_FullScreen_16px.svg')
    expect(data).toMatchSnapshot(true)
  })
})

describe('generateMonochromListObject', () => {
  it('generateMonochromListObject return a list of icons with a component name', () => {
    const data = generateMonochromListObject(iconList)
    expect(data).toMatchSnapshot(true)
    expect(data[0].cat).toEqual('Navigation')
    expect(data[0].name).toEqual('Arrow_Arrow--Top')
    expect(data[0].componentName).toEqual('ArrowArrowTop16')
    expect(data[0].size).toEqual('16px')
  })
})

describe('generateColorListObject', () => {
  it('generateColorListObject return a list of icons without a component name', () => {
    const data = generateColorListObject(iconList)
    expect(data).toMatchSnapshot(true)
    expect(data[0].cat).toEqual('Navigation')
    expect(data[0].name).toEqual('Arrow_Arrow--Top')
    expect(data[0].componentName).toBeUndefined()
    expect(data[0].size).toEqual('16px')
  })
})

describe('checkPath', () => {
  it('should call mkdir to create the directory if it does not exist', () => {
    fs.exists = jest.fn((p, cb) => cb(false))
    fs.mkdir = jest.fn()
    checkPath('foo/bar').then(expect(fs.mkdir).toHaveBeenCalled())
  })

  it('should resolve without problem if the directory exist', () => {
    fs.exists = jest.fn((p, cb) => cb(true))
    fs.mkdir = jest.fn()
    checkPath('foo/bar').then(expect(fs.mkdir).not.toHaveBeenCalled())
  })
})

describe('writeFile', () => {
  it('should call fs.writeFile', () => {
    fs.writeFile = jest.fn()
    writeFile('foo/bar').then(expect(fs.writeFile).toHaveBeenCalled())
  })
})

describe('genrateData', () => {
  fs.exists = jest.fn((p, cb) => cb(true))

  it('should call fs.writeFile and write a valid JSON', () => {
    let fileResult
    fs.writeFile = jest.fn((path, data, utf8, callback) => {
      fileResult = data
      callback(false)
    })
    generateIconsData([[...iconList], [...iconList]]).then(() => {
      expect(fs.writeFile).toHaveBeenCalled()
      expect(fileResult).toMatchSnapshot()
    })
  })

  it('should reject errors if data is invalid', () => {
    fs.writeFile = jest.fn((path, data, utf8, callback) => {
      callback(false)
    })
    generateIconsData([[{ foo: 'bar' }], [{ foo: 'bar' }]]).catch(err => {
      expect(err).toBe(
        `generate icons data : the data received seems to be malformed\nTypeError: Cannot read property 'split' of undefined`
      )
    })
  })

  it('should reject errors if file cannot be saved', () => {
    fs.writeFile = jest.fn((path, data, utf8, callback) => {
      callback('error message mock')
    })
    generateIconsData([[...iconList], [...iconList]]).catch(err => {
      expect(err).toBe(`while trying save the file : error message mock`)
    })
  })
})
