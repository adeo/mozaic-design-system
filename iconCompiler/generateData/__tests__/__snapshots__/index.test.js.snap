// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateColorListObject generateColorListObject return a list of icons without a component name 1`] = `
Array [
  Object {
    "cat": "Navigation",
    "fileName": "Navigation_Arrow_Arrow--Top_16px.svg",
    "name": "Arrow_Arrow--Top",
    "size": "16px",
  },
  Object {
    "cat": "Media",
    "fileName": "Media_FullScreen_24px.svg",
    "name": "FullScreen",
    "size": "24px",
  },
]
`;

exports[`generateMonochromListObject generateMonochromListObject return a list of icons with a component name 1`] = `
Array [
  Object {
    "cat": "Navigation",
    "componentName": "ArrowArrowTop16",
    "fileName": "Navigation_Arrow_Arrow--Top_16px.svg",
    "name": "Arrow_Arrow--Top",
    "size": "16px",
  },
  Object {
    "cat": "Media",
    "componentName": "FullScreen24",
    "fileName": "Media_FullScreen_24px.svg",
    "name": "FullScreen",
    "size": "24px",
  },
]
`;

exports[`genrateData should call fs.writeFile and write a valid JSON 1`] = `
"{
  \\"color\\": [
    {
      \\"cat\\": \\"Navigation\\",
      \\"size\\": \\"16px\\",
      \\"name\\": \\"Arrow_Arrow--Top\\",
      \\"fileName\\": \\"Navigation_Arrow_Arrow--Top_16px.svg\\"
    },
    {
      \\"cat\\": \\"Media\\",
      \\"size\\": \\"24px\\",
      \\"name\\": \\"FullScreen\\",
      \\"fileName\\": \\"Media_FullScreen_24px.svg\\"
    }
  ],
  \\"monochrom\\": [
    {
      \\"cat\\": \\"Navigation\\",
      \\"size\\": \\"16px\\",
      \\"name\\": \\"Arrow_Arrow--Top\\",
      \\"fileName\\": \\"Navigation_Arrow_Arrow--Top_16px.svg\\",
      \\"componentName\\": \\"ArrowArrowTop16\\"
    },
    {
      \\"cat\\": \\"Media\\",
      \\"size\\": \\"24px\\",
      \\"name\\": \\"FullScreen\\",
      \\"fileName\\": \\"Media_FullScreen_24px.svg\\",
      \\"componentName\\": \\"FullScreen24\\"
    }
  ]
}"
`;

exports[`getInfosFromName getsInfosFromName detect category, size, name, and filename from file name 1`] = `
Object {
  "cat": "Media",
  "fileName": "Media_FullScreen_16px.svg",
  "name": "FullScreen",
  "size": "16px",
}
`;
